src/utils/style-utils.js

```diff
@@ -1,139 +1,77 @@
-  if (layersDiff.updates.some(node => node.layer.get('ref'))) {
-    map.setStyle(nextStyle.toJS());
-    return;
+// Prepare a map style object for diffing
+// If immutable - convert to plain object
+// Work around some issues in the styles that would fail Mapbox's diffing
+export function normalizeStyle(style) {
+  if (!style) {
+    return null;
   }
-
-  for (const enter of sourcesDiff.enter) {
-    map.addSource(enter.id, enter.source.toJS());
+  if (typeof style === 'string') {
+    return style;
   }
-  for (const update of sourcesDiff.update) {
-    updateStyleSource(map, update);
+  if (style.toJS) {
+    style = style.toJS();
   }
+  const layerIndex = style.layers.reduce(
+    (accum, current) => Object.assign(accum, {[current.id]: current}),
+    {}
+  );
 
-  for (const exit of layersDiff.exiting) {
-    if (map.style.getLayer(exit.id)) {
-      map.removeLayer(exit.id);
-    }
-  }
-  for (const update of layersDiff.updates) {
-    if (!update.enter) {
-      // This is an old layer that needs to be updated. Remove the old layer
-      // with the same id and add it back again.
-      map.removeLayer(update.id);
-    }
-    map.addLayer(update.layer.toJS(), update.before);
-  }
+  style.layers = style.layers.map(layer => {
+    layer = Object.assign({}, layer);
 
-  for (const exit of sourcesDiff.exit) {
-    map.removeSource(exit.id);
-  }
-}
+    // Breaks style diffing :(
+    delete layer.interactive;
 
-/* eslint-disable max-len */
-function checkForEqualLayerSourceChanges(sourceExit, nextLayers) {
-  const sourceIds = sourceExit.map(s => s.id);
-  const layersNotRemoved = nextLayers.filter(lyr => sourceIds.includes(lyr.get('source')));
-  if (layersNotRemoved.size) {
-    // because of this, no source/layer changes will take effect if there is an error
-    throw new Error(`You must remove any layers associated with sources you are removing: ${layersNotRemoved.map(l => l.get('id')).toJS().join('')}`);
-  }
+    const layerRef = layerIndex[layer.ref];
+    // Style diffing doesn't work with refs so expand them out manually before diffing.
+    if (layerRef) {
+      delete layer.ref;
+      if (layerRef.type !== undefined) {
+        layer.type = layerRef.type;
+      }
+      if (layerRef.source !== undefined) {
+        layer.source = layerRef.source;
+      }
+      if (layerRef['source-layer'] !== undefined) {
+        layer['source-layer'] = layerRef['source-layer'];
+      }
+      if (layerRef.minzoom !== undefined) {
+        layer.minzoom = layerRef.minzoom;
+      }
+      if (layerRef.maxzoom !== undefined) {
+        layer.maxzoom = layerRef.maxzoom;
+      }
+      if (layerRef.filter !== undefined) {
+        layer.filter = layerRef.filter;
+      }
+      if (layerRef.layout !== undefined) {
+        layer.layout = layer.layout || {};
+        layer.layout = Object.assign({}, layer.layout, layerRef.layout);
+      }
+    }
+    return layer;
+  });
+
+  return style;
```

Comment:
Should we drop support for `ref`s? Seems like mapbox dropped support for them almost two years ago:
https://github.com/mapbox/mapbox-gl-js/pull/3486